### PR TITLE
XFA - Add a parser for XFA files

### PR DESCRIPTION
**Pull request by calixteman**
- the parser is base on a class extending XMLParserBase
 - it handle xml namespaces:
   * each namespace is assocated with a builder
   * builder builds nodes belonging to the namespace
   * when a node is inserted in the parent namespace compatibility is checked (if required)
 - to avoid name collision between xml names and object properties, use Symbol.

### CODE REVIEW

File: test/unit/jasmine-boot.js [hunk 1]
```diff
@@ -84,6 +84,7 @@ async function initializePDFJS(callback) {
       "pdfjs-test/unit/unicode_spec.js",
       "pdfjs-test/unit/util_spec.js",
       "pdfjs-test/unit/writer_spec.js",
+      "pdfjs-test/unit/xfa_parser_spec.js",
```
Comment (Snuffleupagus) [inline test/unit/jasmine-boot.js:87]: Unless these unit-tests can't for some reason be run in Node.js, please don't forget to update https://github.com/mozilla/pdf.js/blob/master/test/unit/clitests.json as well :-)

**Comment by calixteman** [comment]
The objects in `config.js` or in `xdp.js` reflect what we've in the documentation, for example the `acrobat` element documentation is:
http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.364.2157&rep=rep1&type=pdf#page=1298&zoom=auto,-207,679
and the corresponding class:
https://github.com/mozilla/pdf.js/pull/12879/files#diff-0fde7327b01926a0a21537877b58611fa916d5cd228e33673678bc37565db84aR26

File: gulpfile.js [hunk 2]
```diff
@@ -1447,6 +1447,7 @@ function buildLib(defines, dir) {
     gulp.src(
       [
         "src/{core,display,shared}/*.js",
+        "src/core/xfa/*.js",
```
Comment (Snuffleupagus) [inline gulpfile.js:1450]: I haven't tested, but wouldn't it suffice to simply do:
```suggestion
        "src/{core,display,shared}/**/*.js",
```

File: src/core/xfa/builder.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import { UnknownNamespace } from "./unknown.js";
+import { warn } from "../../shared/util.js";
+
+class Root extends XFAObject {
+  constructor() {
+    super(-1, "root", Object.create(null));
+    this.element = null;
+  }
+
+  [$onChild](child) {
```
Comment (timvandermeij) [inline src/core/xfa/builder.js:28]: How does this syntax work? It doesn't look like a valid function name with these special characters in it.

Comment (calixteman) [reply src/core/xfa/builder.js:28]: `$onChild` is defined in `xfa_object` like this: `const $onChild = Symbol();`
And the way to use this symbol as a function name is to use brackets:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol#symbol_wrapper_objects_as_property_keys
The only way to call this function is to have the symbol so there's no access by name (as a string).

Comment (timvandermeij) [reply src/core/xfa/builder.js:28]: Cool, learned something new then :-)

It's not entirely clear yet as to why this is needed exactly (i.e., why access by name causes conflicts here), but that requires a more in-depth look. This syntax was just something I noticed and didn't know; thanks for explaining this!

Comment (calixteman) [reply src/core/xfa/builder.js:28]: In the specs there is something call SOM expressions to select nodes:
http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.364.2157&rep=rep1&type=pdf#page=101&zoom=auto,-207,766
These expressions seem to be used in different places.
And so in order to make easy their interpretation, each part of the expression can be searched as a own property of the object without the need to keep a flag somewhere to check if the property exists by spec or if it is an implementation detail.
And I guess it'll avoid any bad use an attacker could do with specific expressions.
So my idea was to have objects which are the exact reflection of specs in hiding implementations details and thx to that we can enjoy the object model stuff like overriding `$onChildCheck` method when a specific node can accept a node from another namespace.
So from my pov, it makes the implementation safer, simpler and reduce memory use (no need to track spec properties...).
And it's unusual I added this `$` symbols to help to see them.

File: src/core/xfa/xfa_object.js [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+      builder.clean(this[$cleanup]);
+      delete this[$cleanup];
+    }
+  }
+
+  [$isTransparent]() {
+    return this.name === "";
+  }
+
+  [$dump](buf = [], level = 0) {
```
Comment (brendandahl) [inline src/core/xfa/xfa_object.js:84]: Is there an advantage to using our own output format vs json stringifying an object?

File: src/core/xfa/xfa_object.js [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+class Option10 extends IntegerObject {
+  constructor(nsId, name) {
+    super(nsId, name, 1, n => n === 0);
+  }
+}
+
+export {
+  $clean,
```
Comment (brendandahl) [inline src/core/xfa/xfa_object.js:234]: Are there going to be many more of these? I like the idea of keeping things separate using symbols, but having to export every method seems very awkward vs prefxing with a `$` and ensuring those properties aren't overwritten.

**Review by brendandahl — CHANGES_REQUESTED** [review]

File: src/core/xfa/xfa_object.js (hunk 5 above)
Comment (calixteman) [reply src/core/xfa/xfa_object.js:234]: The other advantage of having a `Symbol` is that they aren't returned by `Object.getOwnPropertyNames` and so it avoid to check if name starts with a `$` when evaluating a SOM expression:
http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.364.2157&rep=rep1&type=pdf#page=86&zoom=auto,-207,138
Anyway, you're right: the drawback that we need to import symbols.

File: src/core/xfa/xfa_object.js (hunk 4 above)
Comment (calixteman) [reply src/core/xfa/xfa_object.js:84]: I just discover `toJSON`: I'll update the patch with.
Thx for the tip.

File: src/core/xfa/xfa_object.js (hunk 5 above)
Comment (brendandahl) [reply src/core/xfa/xfa_object.js:234]: As discussed, we'll go with symbol since it makes SOM stuff easier.

**Review by brendandahl — APPROVED** [review]

**Pull request merged**